code/modules/food_and_drink/cakes.dm

```diff
@@ -24,62 +24,482 @@
+
+		var/list/returns = build_cake(s,user,2,layer_tag,replacetext)
+		slices = returns[1]
+		slice_candle = returns[2]
+
+		var/transferamount = (src.amount/src.clayer)/slices //amount of reagent to transfer to slices
+		for(var/i=1,i<=slices,i++) //generating child slices of the parent template
+			var/obj/item/reagent_containers/food/snacks/cake/custom/schild = new /obj/item/reagent_containers/food/snacks/cake/custom
+			schild.icon_state = "slice-overlay"
+			for(var/i2=1,i2<=s.overlays.len,i2++) //looping through parent overlays and copying them over to the children
+				schild.UpdateOverlays(s.GetOverlayImage("[s.overlay_refs[i2]]"),"[s.overlay_refs[i2]]")
+			if(slice_candle) //making sure there's only one candle :)
+				if(slice_candle == 1)
+					schild.UpdateOverlays(new /image('icons/obj/foodNdrink/food_dessert.dmi',"slice-candle"),"slice-candle")
+				if(slice_candle == 2)
+					schild.UpdateOverlays(new /image('icons/obj/foodNdrink/food_dessert.dmi',"slice-candle_lit"),"slice-candle_lit")
+				slice_candle = 0
+				cake_candle = 0
+				schild.cake_candle = 1
+				if(src.litfam) //light update
+					src.put_out()
+					schild.ignite()
+			src.reagents.trans_to(schild,transferamount) //setting up the other properties of the slice
+			schild.pixel_x = rand(-6, 6)
+			schild.pixel_y = rand(-6, 6)
+			for(var/b=1,b<=src.food_effects.len,b++)
+				if(src.food_effects[b] in schild.food_effects)
+					continue
+				schild.food_effects += src.food_effects[b]
+			schild.w_class = 1
+			schild.quality = src.quality
+			schild.name = "slice of [src.name]"
+			schild.desc = "a delicious slice of cake!"
+			schild.food_color = src.food_color
+			schild.sliced = 1
+			schild.amount = 1
+
+			schild.set_loc(get_turf(src.loc))
+		qdel(s) //cleaning up the template slice
+		if(src.clayer == 1) //qdel(src) if there was only one layer to the cake, otherwise, decrement the layer
+			qdel(src)
+		else
+			src.clayer--
+
+
+	proc/stack_cake(var/obj/item/reagent_containers/food/snacks/cake/custom/c,var/mob/user)
+		if(!(src.clayer<3))
+			return
+		if(c.clayer>=3)
+			return
+		if(c.sliced)
+			return
+
+		user.u_equip(c)
+		c.set_loc(user)
+		src.clayer++
+		src.reagents.maximum_volume += 100
+		src.amount += 10
+		c.reagents.trans_to(src,c.reagents.total_volume)
+
+		for(var/i=1,i<=c.food_effects.len,i++) //adding food effects to the src that arent already present
+			if(c.food_effects[i] in src.food_effects)
+				continue
+			src.food_effects += c.food_effects[i]
+
+		var/singlecake //logging the clayer before changing it later
+		if(c.clayer == 1)
+			singlecake = 1
+
+		for(var/i=1,i<=src.overlays.len,i++) //looking for candles and if you set a cake on top of it, it pops off!
+			if(("[src.overlay_refs[i]]" == "cake[(src.clayer)-1]-candle") || ("[src.overlay_refs[i]]" == "cake[(src.clayer)-1]-candle_lit") || ("[src.overlay_refs[i]]" == "cake[(src.clayer)-2]-candle") || ("[src.overlay_refs[i]]" == "cake[(src.clayer)-2]-candle_lit"))
```

Comment:
Iterating over overlays like that is not a great solution. Iterating over the list of something like `list("cake[(src.clayer)-1]-candle", "cake[(src.clayer)-1]-candle_lit" ...)` would seem cleaner. Even better solution would be to have the proper information stored somewhere else and not just in the overlays. `cake_candle` already tells if if you have a candle, right? You could also have a var that tells you what's its overlay slot name. Or have the slot name be directly in `cake_candle`.

Reply:
could also be associated ye